### ISSUE TITLE
Clarify/document hotp-verification output for 6 Admin/User PINS and firmware version

### ISSUE THREAD

**Issue by tlaurion** [Nitrokey/nitrokey-hotp-verification#36]
Wrongly reported under https://github.com/linuxboot/heads/issues/1726

My answer at https://github.com/linuxboot/heads/issues/1726#issuecomment-2241140702

@nestire answer (incomplete) at https://github.com/linuxboot/heads/issues/1726#issuecomment-2243048271

--

End user asked clarifications under Dasharo Premier support at (answered by me): https://matrix.to/#/!RNcjJXCGHiyxXCHpKv:matrix.org/$A9IbvnjLuw1EaXAynHtvlrnewrkPqYpMA2kae5t1ONM?via=matrix.org&via=nitro.chat&via=matrix.3mdeb.com

**Comment by tlaurion** [comment]
User said under Matrix

> I've just read the seal-hotpkey source code and it suggests that this is expected behaviour for a month, but you are encouraged to change it - will do that, thanks :-)

@jans23 @nestire @daringer : seal-hotpkey get the output of gpg which in this case is a bit irrelevant since NK3 implement HOTP secret independently of GPG. The "age" being <1 old is still valid, where card counters don't apply.

So technically, logic applied under seal-hotp-key to check counters is invalid as of today.

**Comment by daringer** [comment]
It looks like the output of `hotp_verification info` is only used inside the lower part of the referred code snippet. Essentially `hotp_token_info` is parsed for the admin pin retry count.

As `hotp_verification` is anyways only used in HEADS, wouldn't it make sense to tailor its output more towards the needs of HEADS? How about you propose what `hotp_verification info` should output (best case with an example) for the different security tokens accepted (should be: nk-pro/storage, librem, nk3) and we adapt its output accordingly?

Going this way the entire `hotp_token_info` parsing could be removed from HEADS.

**Comment by tlaurion** [comment]
@daringer : could hotp-verification
- output nk3 firmware version, not the secret secure element version?
- could hotp-verification report on HOTP/GPG Admin/User PIN?
- Could hotp-verification make nk1/2 pro/librem key/nk3 usage a total abstraction of the HOTP sealing firmware requirements?

**Comment by daringer** [comment]
> output nk3 firmware version, not the secret secure element version?

will check

> could hotp-verification report on HOTP/GPG Admin/User PIN?

I suppose you mean it should report remaining tries, if so: will check

> Could hotp-verification make nk1/2 pro/librem key/nk3 usage a total abstraction of the HOTP sealing firmware requirements?

Don't understand this, can you elaborate what you mean here? Also please keep in mind we are talking about `hotp_verification info` here. 

Generally, an example of the expected output of `hotp_verification info` would help to get the same understanding.

**Comment by daringer** [comment]
ok, we can do 1) + 2) without too much hassle, @tlaurion could you please tell (note down an example) how you would like the `hotp_verification info` output to look like ?

**Comment by tlaurion** [comment]
> The first and second points should not be too complicated, @tlaurion could you please tell (note down an example) how you would like the `hotp_verification info` output to look like ? 

Sorry I haven't looked into this.
Well I do not know what is available there. My point was to make this abstract of nk1 nk2nk3 lk2, Heads should depend on hotp-verification to give user feedback that can be understandable by cognizant beings.

As of now the output doesn't make any sense providing info that are not even in releases notes so cannot even be crosslinked.

I already referred to what code of heads do and where things could be adapted depending of what hotp-verification could provide.

For the rest this is nitrokey problem, not mine I'm sorry

**Comment by tlaurion** [comment]
@daringer : maybe we have to flip this around and see what Heads official docs currently says at https://github.com/linuxboot/heads-wiki/blob/master/Installing-and-Configuring/configuring-keys.md rendered at https://osresearch.net/Configuring-Keys/#oem-factory-resetre-ownership

Thinking points:
- hotp-verification was an implementation detail. With NK3, GPG Admin pins has nothing to do with HOTP. How would you word that?
- OEM Factory reset doesn't set it, its set on first use. What would be the guidelines for HOTP related PIN, how users are supposed to change it etc? what are the official Nitrokey Docs, how can end user make sense of it all?


On hotp-verification output:
- Please output nk2 output, detail its meaning
- Please output nk3 output, detail its meaning

Let's start from there @daringer ?

**Comment by tlaurion** [comment]
@daringer do you think we can do a crunch here and fix this issue prior of 2024-11-20 feature freeze of Heads at https://github.com/linuxboot/heads/issues/1821 ?

I had again an issue at https://github.com/Nitrokey/nitrokey-hotp-verification/issues/37#issuecomment-2462924753 related to hotp-verification not reporting the good thing and therefore Heads doing wrong assumptions.

Eg:
![signal-2024-11-07-131256](https://github.com/user-attachments/assets/261f6ff3-f853-488f-b587-1cea02c54966)
Message: `Not trying default PIN (12345678) only 0 attempt left` si to say the least misleading

**Comment by tlaurion** [comment]
This discussion needs answers to https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36#issuecomment-2284078511 to follow through @jans23

**Comment by JonathonHall-Purism** [comment]
@tlaurion Could you paste / screenshot the result of `hotp_verification info` in this situation?

It's looking for `^\s*Card counters: Admin (\d),.*$`.  If the output changed, we can change the regex in Heads.

**Comment by tlaurion** [comment]
Repro:


### nk3a-mini
test
```
user@heads-tests-deb12-nix:~/heads$ nitropy nk3 test
Command line tool to interact with Nitrokey devices 0.4.47
Found 1 Nitrokey 3 device(s):
- Nitrokey 3 at /dev/hidraw0

Running tests for Nitrokey 3 at /dev/hidraw0

[1/5]	uuid     	UUID query              	SUCCESS  	EF25D848139028D30000000000000000
[2/5]	version  	Firmware version query  	SUCCESS  	v1.7.2
[3/5]	status   	Device status           	SUCCESS  	Status(init_status=<InitStatus: 0>, ifs_blocks=238, efs_blocks=465, variant=<Variant.NRF52: 2>)
Running SE050 test: |                                                                                                                                                                                              
[4/5]	se050    	SE050                   	SUCCESS  	SE050 firmware version: 3.1.1 - 1.11, (persistent: (31432,), transient_deselect: (607,), transient_reset: (592,))
Please press the touch button on the device ...
Please press the touch button on the device ...
[5/5]	fido2    	FIDO2                   	SUCCESS  	

5 tests, 5 successful, 0 skipped, 0 failed

Summary: 1 device(s) tested, 1 successful, 0 failed

```

reset:
```
user@heads-tests-deb12-nix:~/heads$ nitropy nk3 secrets reset
Command line tool to interact with Nitrokey devices 0.4.47
Do you want to continue? [y/N]: y
Please touch the device if it blinks
Done
```

Then from Heads:
```
$ cat hotp_verification_info-output_nk3a_mini.log 
HOTP code verification application, version 1.6
Connected device status:
	Card serial: 0xEF25D848
	Firmware: v4.13
	Card counters: PIN is not set - set PIN before the first use
Operation success

```

### nk3-nfc

test:
```
user@heads-tests-deb12-nix:~/heads$ nitropy nk3 test
Command line tool to interact with Nitrokey devices 0.4.47
Found 1 Nitrokey 3 device(s):
- Nitrokey 3 at /dev/hidraw0

Running tests for Nitrokey 3 at /dev/hidraw0

[1/5]	uuid     	UUID query              	SUCCESS  	7BE66C6C09655959911E4A5958996AEF
[2/5]	version  	Firmware version query  	SUCCESS  	v1.7.2
[3/5]	status   	Device status           	SUCCESS  	Status(init_status=<InitStatus: 0>, ifs_blocks=41, efs_blocks=462, variant=<Variant.LPC55: 1>)
Running SE050 test: |                                                                                                                                                                                              
[4/5]	se050    	SE050                   	SUCCESS  	SE050 firmware version: 3.1.1 - 1.11, (persistent: (32767,), transient_deselect: (191,), transient_reset: (176,))
Please press the touch button on the device ...
Please press the touch button on the device ...
[5/5]	fido2    	FIDO2                   	SUCCESS  	

5 tests, 5 successful, 0 skipped, 0 failed

Summary: 1 device(s) tested, 1 successful, 0 failed

```
reset
```
user@heads-tests-deb12-nix:~/heads$ nitropy nk3 secrets reset
Command line tool to interact with Nitrokey devices 0.4.47
Do you want to continue? [y/N]: y
Please touch the device if it blinks
Done
```

Heads collected ouput:
```
user@heads-tests-deb12-nix:~/heads$ cat /media/hotp_verification_info-output_nk3_nfc.log 
HOTP code verification application, version 1.6
Connected device status:
	Card serial: 0x7BE66C6C
	Firmware: v4.13
	Card counters: PIN is not set - set PIN before the first use
Operation success
```

@JonathonHall-Purism yes, we could change regex (must: at least for https://github.com/linuxboot/heads/issues/1821), even more if downstream plans to do one Heads release a year, maybe two).
@daringer  that won't change the fact that the output reported by hotp_verification is totally irrelevant (UX? Scripts not working today in lack of proper output.)
- Firmware 4.13? 
- How to get counts of HOTP bad attempts?
- What to change in code so that GPG Admin PIN is not thought to set HOTP as for Nk2/Librem keys?

**Comment by JonathonHall-Purism** [comment]
Relevant line (same in both): `	Card counters: PIN is not set - set PIN before the first use`

It sounds like we are doing the right thing for the wrong reason.  If no PIN is set, it doesn't make sense to try the default, but our message could be more precise, just by detecting this text.

I don't have an NK3.  Is it possible to seal HOTP in this state?  What PIN is the user supposed to enter?  Based on that, what should seal-hotpkey actually do?

**Comment by tlaurion** [comment]
> Relevant line (same in both): ` Card counters: PIN is not set - set PIN before the first use`
> 
> It sounds like we are doing the right thing for the wrong reason. If no PIN is set, it doesn't make sense to try the default, but our message could be more precise, just by detecting this text.

Would be better but still really problematic:
> 
> I don't have an NK3. Is it possible to seal HOTP in this state? What PIN is the user supposed to enter? Based on that, what should seal-hotpkey actually do?

The 'secret' app gets its 'GPG Admin PIN" (Wrong naming...) set on first HOTP sealing. Exactly there (as of today, after reboot past oem-factory-reset run, which seals TPMTOTP secret through reverse HOTP in nk3 into its secret app):
quote from https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36#issuecomment-2462929736:

> ![signal-2024-11-07-131256](https://github.com/user-attachments/assets/261f6ff3-f853-488f-b587-1cea02c54966)
> Message: `Not trying default PIN (12345678) only 0 attempt left` si to say the least misleading

But this hides another more important problem, as noted under https://github.com/Nitrokey/nitrokey-hotp-verification/issues/37#issue-2641745207, quoting:
> - Resetting TPM/sealing TOTP through reverse HOTP fails
> 
> Screenshot:
> ![signal-2024-11-07-123429](https://github.com/user-attachments/assets/7ec0da75-e60f-4ac4-b742-d4c93b733588)
> 

This means....
- No user re-ownership can happen if OEM seals HOTP secret currenlty (without needing a second computer with nitropy to reset secret)
- Meaning that currently, HOTP isn't/can't be sealed by OEM???


We need:
- [x] hotp-verification reporting properly on nk3 secrets Admin count (reminder, we have GPG Admin Pin, GPG User PIN and now Nk3 secret Admin PIN, where heads is aware/deals and report only of GPG PINs and can only set once HOTP related Admin PIN, but cannot reset it)
- [x] hotp-verification reporting properly on firmware meta version (expected here: 1.7.2. I reask: what is 4.13?!?! And what is the end user supposed to do with that information?)
- [x] How to get counts of HOTP bad attempts?
- [x] What to change in code so that GPG Admin PIN is not thought to set HOTP (secret app Admin PIN) as for Nk2/Librem keys?
- [ ] And most importantly, a mechanism to change that secret app passphrase/reset it just like for nk2/librem keys (which were GPG Admin PIN). We cannot expect users to use nitropy on a second computer, can we? Ohterwise no re-ownership possible with TPMTOPT reversed seal in HOTP.

Thoughts and plan of action @daringer @jans23 @JonathonHall-Purism @sosthene-nitrokey @szszszsz?

**Comment by JonathonHall-Purism** [comment]
To make the user re-ownership flow work:
* surely we could implement the secrets PIN reset performed by nitropy in C or something, right?  I don't know how this works and cannot really volunteer to do it as we're not selling NK3, but it does not seem this should fundamentally require python
    * then OEM reset could do this for nk3
* could OEM reset also perform a dummy HOTP seal for nk3, just to set the initial secrets PIN?  E.g. seal it with 0000000.... or something, next boot will ask to reseal anyway, but the PIN will be set now

If the PIN is set, do we get a valid attempt counter?

**Comment by JonathonHall-Purism** [comment]
>... so no need to sail it with a dummy PIN. But there is a need for reset/passphrase change that is for sure

I mean do a dummy HOTP seal with the _real_ pin, but dummy HOTP secret of all-zeros or something.  Effectively, just setting the secrets PIN.

(Though a proper mechanism to just set the PIN would be much better.)

If we eventually have forward-sealing, we could forward-seal at that time with the real PIN and real HOTP secret.

**Comment by tlaurion** [comment]
Just did a test to see which counter (Admin/User) decrease when failing HOTP sealing. 

Surprise surprise, both are decremented on bad TPMTOTP reverse HOTP sealing over nk3?
```
cat /media/hotp_verification_info-output_nk3a_mini-secret_admin_pin_set-bad_hotp_reseal.txt 
HOTP code verification application, version 1.6
Connected device status:
	Card serial: 0x7BE66C6C
	Firmware: v4.13
	Card counters: Admin 6, User 6
Operation success
```

**Comment by tlaurion** [comment]
@daringer @jans23 @sosthene-nitrokey @szszszsz: ping.

Please let me know that you have clear understanding of the problems at stake here and detailed, with screenshots, under https://github.com/Nitrokey/nitrokey-hotp-verification/issues/36#issuecomment-2463012987

TLDR: 
- once HOTP is sealed the first time, accepting there any PIN given to set what is called today the `Nitrokey Admin PIN`, this PIN cannot be changed unless, as screenshot show, running 'nitropy nk3 secret reset' or using 'Nitrokey App 2'.
- This means that no re-ownership is possible
- This means that OEM factory reset needs physical presence and blocks feature requests of unattended, randomized secret provisioning and re-ownership
- Version of firmware '4.13'
- Admin and User 'Nitrokey PINs' decrement at the same time at HOTP sealing failing.

**I cannot highlight enough that 'Nitrokey Admin PIN' can only be configured once, and does not warn that there is a PIN being set when it does so**

**Comment by daringer** [comment]
please define priorities for the 4 newly created issues @tlaurion here - what needs to be there for a release (feature freeze) what doesn't.

**Comment by tlaurion** [comment]
For feature freeze, we need re-ownership to work as intended so that OEM can seal HOTP and end user can define new secrets per re-ownership.

This implies, in order or priority:
- Allow factory reset (no change PIN) without user presence #41. So that Heads can reset secure element alongside gpg factory reset and expect to set GPG Admin PIN for Secret Element Admin PIN (User PIN/Admin PIN duplicated count output seems irrelevant here, not sure how to present proper UX without confusing user but that is #39). This is my blocker.
- Make info Output more verbose/consistent for admin/user pin Counts (#38) to stop confusing users.
- Add cli option to request secret app admin/user pin counts (#39): on this, for nk2/librem key, discussions with @JonathonHall-Purism are needed too make sure of no regression in UI
- Add option to change the nk3 secret apps pin for Re-Onwership #40. This would be nice to have, but not sure needed if we can reset secret app. A re-ownership as up to now was resetting the whole security dongle (nk2/librem key), the arrival ok nk3 seperated things. I cannot tell what is best for other nk3 use cases, only thing I can say here is that since hotp-verification is silently setting secret app admin/user PIN on first HOTP seal, if the PIN there is entered with a typo, the user will be confused (just like I was when I discovered this) and won't be able to change it. He couldn't change it anyway because typo when asked to enter old PIN. With GPG, changing/unpocking User PIN can be done with Admin PIN, and if Admin PIN is locked, he can reset Admin PIN with "recovery key" if setup (not done by default). Since you said secret app User/Admin PIN is the same, I do not see this being really useful as opposed to a secret app reset where hotp-verification should prompt for PIN and wipe secret app, which is #41 


So in order of priority: 
- #41 (blocker for oem-factory-reset needed changes for OEM factory reset and now impossible re-ownership because use cannot change HOTP secret and need to know OEM HOTP secret: this is bad, and then needs second computer. 
- Then #38:  hotp-verification output being usable by Heads  (firmware version :do i need to upgrade)
- Then #39 (so end-user can understand output:, Single PIN/User/Admin PIN: no regression for nk2/librem keys and heads)

I do not think #40 is needed, but for users that know their HOTP PIN to change it before going abroad, or when they think someone saw them type HOTP sealing PIN on post firmware upgrade (wanted: at firmware upgrade), just so par with nk2/librem key which could be done through gpg before and now can't.

TLDR: #41>#38>#39, nice to have #40

**Comment by daringer** [comment]
okok, yeah #41 will not be possible, currently I do not support this overall, we will discuss this, but generally it is currently not an option. Also, given the firmware nature of this, this is independent to HEADS feature freeze.

So we'll focus on #38 and #39 .... please make sure all asked information to finalize the work is stated in the respective tickets

**Comment by tlaurion** [comment]
> okok, yeah #41 will not be possible, currently I do not support this overall, we will discuss this, but generally it is currently not an option. Also, given the firmware nature of this, this is independent to HEADS feature freeze. Even if I would greenlight this, a full firmware release for the nitrokey 3 takes time and is nothing to be pushed through within a week.
> 
> So we'll focus on #38 and #39 .... please make sure all asked information to finalize the work is stated in the respective tickets

Then #41 needs to be spitted into physical presence removal (firmware: later) and hotp-verification being able to do équivalent of nitropy nk3 secret reset (needed). Please open missing issue. 

Commented at https://github.com/Nitrokey/nitrokey-hotp-verification/issues/41#issuecomment-2481898790

**Comment by tlaurion** [comment]
@sosthene-nitrokey just want to make sure we are at the same page:

> <nk3
> ```
> HOTP code verification application, version 1.7
> Connected device status:
> Card serial: 0x7BE66C6C
>          Firmware Nitrokey 2: v1.7.1
> OpenPGP smartcard PIN counters : Admin: 3, User: 3
> Operation success
> ```
> 
> nk3:
> ```
> HOTP code verification application, version 1.7
> Connected device status:
> Card serial: 0x7BE66C6C
>          Firmware Nitrokey 3: v1.7.1
>          Firmware Secrets app: v4.13
> Secret app PIN counter : 6
> OpenPGP smartcard PIN counters : Admin: 3, User: 3
> Operation success
> ```

_Originally posted by @tlaurion in https://github.com/Nitrokey/nitrokey-hotp-verification/issues/38#issuecomment-2487450702_

**Comment by tlaurion** [comment]
> For feature freeze, we need re-ownership to work as intended so that OEM can seal HOTP and end user can define new secrets per re-ownership.
 

> TLDR: #41>#38>#39, nice to have #40

#41 was deemed impossible because physical presence removal required nk3 firmware change, and led to #42

So priorities are:

#42>#41>#38>#39, nice to have #40
